### PR TITLE
TST: Add additional test for future warning when call Series.str.cat(Series.str)

### LINKED ISSUE #28277: `series.str.cat(series.str)` is concatenating only the largest string
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

arr = ["AbC", "de", "FGHI", "j", "kLLLm"]

ps = pd.Series(arr)
expect = ps.str.cat(others=ps.str)
print(expect)
Out[16]:
0           NaN
1           NaN
2           NaN
3           NaN
4    kLLLmkLLLm
dtype: object
```

#### Problem description
`series.str.cat(series.str)` is concatenating only the largest string in the series but it should concatenate all the strings element wise.

#### Expected Output
```python
Out[18]:
0        AbCAbC
1          dede
2      FGHIFGHI
3            jj
4    kLLLmkLLLm
dtype: object
```

#### Output of ``pd.show_versions()``

<details>

In [2]: pandas.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-862.14.4.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: None
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 5.1.2
pip: 19.2.3
setuptools: 41.2.0
Cython: 0.29.13
numpy: 1.17.1
scipy: None
pyarrow: 0.14.1
xarray: None
IPython: 7.8.0
sphinx: 2.2.0
patsy: None
dateutil: 2.8.0
pytz: 2019.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10.1
s3fs: 0.3.4
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by xr-chen**
- [x] closes #28277 (Replace xxxx with the Github issue number)
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/strings/test_cat.py [hunk 1]
```diff
@@ -376,3 +376,11 @@ def test_cat_different_classes(klass):
     result = s.str.cat(klass(["x", "y", "z"]))
     expected = Series(["ax", "by", "cz"])
     tm.assert_series_equal(result, expected)
+
+
+def test_cat_on_series_dot_str():
+    ps = Series(["AbC", "de", "FGHI", "j", "kLLLm"])
+
+    res = ps.str.cat(others=ps.str)
```
Comment (mroeschke) [inline pandas/tests/strings/test_cat.py:384]: Like the original issue mentioned https://github.com/pandas-dev/pandas/issues/28277#issuecomment-527851842=, `ps.str.cat(others=ps.str)` should raise an exception here. `ps.str.cat(others=ps)` should be okay

**Review by mroeschke — CHANGES_REQUESTED** [review]

File: pandas/tests/strings/test_cat.py (hunk 1 above)
Comment (xr-chen) [reply pandas/tests/strings/test_cat.py:384]: Oh right, I was misled by the discussion in #29909. Should we fix this issue independently or wait until `Series.str.__iter__` was deprecated?

Comment (mroeschke) [reply pandas/tests/strings/test_cat.py:384]: Can you determine what the behavior when `Series.str.__iter__` will be removed? i.e. will a sensible error raise if `Series.str.__iter__` is removed?

Comment (mroeschke) [reply pandas/tests/strings/test_cat.py:384]: Right. `Series.str.__iter__` is already deprecated and will be removed in 2.0.

I'm asking _if_ you remove `Series.str.__iter__` locally and run the test case again, would a sensible error raise or would it make sense to do some input validation in `cat` to raise a better error message.

Comment (mroeschke) [reply pandas/tests/strings/test_cat.py:384]: Okay then in that case, could you just add a unit test in this PR? The unit test could exercise that the warning of `Series.str.__iter__` when `ps.str.cat(others=ps)` is run, but also have commented out code that would test the `TypeError` exception when `Series.str.__iter__` is removed

Comment (xr-chen) [reply pandas/tests/strings/test_cat.py]: No problem.

File: pandas/tests/strings/test_cat.py [hunk 2]
```diff
@@ -376,3 +376,22 @@ def test_cat_different_classes(klass):
...
     tm.assert_series_equal(result, expected)
+
+
+def test_cat_on_series_dot_str():
+    # GH 28277
+    # Test future warning of `Series.str.__iter__`
+    ps = Series(["AbC", "de", "FGHI", "j", "kLLLm"])
+    with tm.assert_produces_warning(FutureWarning):
+        ps.str.cat(others=ps.str)
+    # The following code will be uncommented if `Series.str.__iter__` is removed.
```
Comment (mroeschke) [inline pandas/tests/strings/test_cat.py:387]:
```suggestion
    # TODO(2.0): The following code can be uncommented when `Series.str.__iter__` is removed.
```

File: pandas/tests/strings/test_cat.py [hunk 3]
```diff
@@ -376,3 +376,22 @@ def test_cat_different_classes(klass):
...
+    """
+    message = re.escape(
```
Comment (mroeschke) [inline pandas/tests/strings/test_cat.py:389]: Could you use `#` instead of a `"""` block for commenting?

**Review by mroeschke — APPROVED** [review]

**Pull request merged**